server/use_cases/add_service_commitments.py

```diff
@@ -39,52 +41,98 @@ def add_service_commitments(commitments_repo, shifts_repo, commitments):
         else:
             valid_commitments.append(commitment)
             valid_indexes.append(i)
-            results.append(None)  # placeholder for now
-
+            results.append(None)
     if not valid_commitments:
         return results
-
-    # check for time overlap in valid shifts
+    # Get user ID from first commitment
+    user_id = valid_commitments[0].volunteer_id
+    # Fetch user's existing commitments
+    try:
+        existing_user_commitments = commitments_repo.fetch_service_commitments(
+            user_id=user_id)
+        if existing_user_commitments is None:
+            existing_user_commitments = []
+    except (AttributeError, TypeError):
+        existing_user_commitments = []
+    existing_commitment_shift_ids = [str(c.service_shift_id)
+                                    for c in existing_user_commitments
+                                    if c.service_shift_id is not None]
+    # Get existing shifts
+    existing_user_shifts = []
+    if existing_commitment_shift_ids:
+        existing_user_shifts = shifts_repo.get_shifts(
+            existing_commitment_shift_ids)
+    # Get valid shifts for current commitments
     valid_shifts = [
-        existing_shifts[c.service_shift_id]
-        for c in valid_commitments
-    ]
-    if check_time_overlap(valid_shifts):
-        for i in valid_indexes:
-            results[i] = {
+        existing_shifts[c.service_shift_id] for c in valid_commitments]
+    final_valid_commitments = []
+    final_valid_indexes = []
+    for i, idx in enumerate(valid_indexes):
+        current_shift = valid_shifts[i]
+        has_overlap = False
+        # Check for overlap with existing user shifts
+        for existing_shift in existing_user_shifts:
+            if (current_shift.shift_start < existing_shift.shift_end and
+                current_shift.shift_end > existing_shift.shift_start):
+                has_overlap = True
+                break
```

Comment:
What happened to calling check_time_overlap? You used to call it here, but now you duplicated the code that's in check_time_overlap in this function.